Unclear error when on mismatched type parameter with a default value

### Code

```Rust
pub struct LooooooongObscureType<A, B>(A, B);
pub struct WrapperWithMismatchedParameters<W = usize, B = Vec<W>>(W, B);

pub struct MainType<H, L> {
    g: H,
    l: L,
}

pub fn f(
    x: MainType<
        LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
        WrapperWithMismatchedParameters<usize, &'static [usize]>,
    >,
) {
    g(x);
}

pub fn g(
    x: MainType<
        LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
        WrapperWithMismatchedParameters,
    >,
) {
    unimplemented!();
}
```


### Current output

```Shell
error[E0308]: mismatched types
  --> src/lib.rs:15:7
   |
15 |     g(x);
   |     - ^ expected `Vec<usize>`, found `&'static [usize]`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `MainType<_, WrapperWithMismatchedParameters<_>>` (`Vec<usize>`)
              found struct `MainType<_, WrapperWithMismatchedParameters<_>>` (`&'static [usize]`)
note: function defined here
  --> src/lib.rs:18:8
   |
18 |   pub fn g(
   |          ^
19 | /     x: MainType<
20 | |         LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
21 | |         WrapperWithMismatchedParameters,
22 | |     >,
   | |_____-

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to 1 previous error
```


### Desired output

```Shell
error[E0308]: mismatched types
  --> src/lib.rs:15:7
   |
15 |     g(x);
   |     - ^ expected `Vec<usize>`, found `&'static [usize]`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `MainType<_, WrapperWithMismatchedParameters<_, Vec<usize>>>`
              found struct `MainType<_, WrapperWithMismatchedParameters<_, &'static [usize]>>`
note: function defined here
  --> src/lib.rs:18:8
   |
18 |   pub fn g(
   |          ^
19 | /     x: MainType<
20 | |         LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
21 | |         WrapperWithMismatchedParameters,
22 | |     >,
   | |_____-

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to 1 previous error
```

or

```
error[E0308]: mismatched types
  --> src/lib.rs:15:7
   |
15 |     g(x);
   |     - ^ expected `Vec<usize>`, found `&'static [usize]`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `MainType<_, WrapperWithMismatchedParameters<_, Vec<usize>>>`
              found struct `MainType<_, WrapperWithMismatchedParameters<_, &'static [usize]>>`
note: function defined here
  --> src/lib.rs:18:8
   |
18 |   pub fn g(
   |          ^
19 | /     x: MainType<
20 | |         LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
21 | |         WrapperWithMismatchedParameters<usize, Vec<usize>>,
22 | |     >,
   | |_____-

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to 1 previous error
```

### Rationale and extra context

The current error message does not tell me the mismatch is in `WrapperWithMismatchedParameters` rather than in `LooooooongObscureType`. It is also inconsistent with the error message I would get if `WrapperWithMismatchedParameters` has without the default types.

### Other cases

When removing the default for the first parameter (which is not the mismatched one) like this:

```Rust
pub struct LooooooongObscureType<A, B>(A, B);
pub struct WrapperWithMismatchedParameters<W, B = Vec<W>>(W, B);

pub struct MainType<H, L> {
    g: H,
    l: L,
}

pub fn f(
    x: MainType<
        LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
        WrapperWithMismatchedParameters<usize, &'static [usize]>,
    >,
) {
    g(x);
}

pub fn g(
    x: MainType<
        LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
        WrapperWithMismatchedParameters<usize>,
    >,
) {
    unimplemented!();
}
```

the error message is clearer:

```
error[E0308]: mismatched types
  --> src/lib.rs:15:7
   |
15 |     g(x);
   |     - ^ expected `Vec<usize>`, found `&'static [usize]`
   |     |
   |     arguments to this function are incorrect
   |
   = note: expected struct `MainType<_, WrapperWithMismatchedParameters<_, Vec<usize>>>`
              found struct `MainType<_, WrapperWithMismatchedParameters<_, &'static [usize]>>`
note: function defined here
  --> src/lib.rs:18:8
   |
18 |   pub fn g(
   |          ^
19 | /     x: MainType<
20 | |         LooooooongObscureType<Vec<&'static [usize]>, &'static [u64]>,
21 | |         WrapperWithMismatchedParameters<usize>,
22 | |     >,
   | |_____-

For more information about this error, try `rustc --explain E0308`.
error: could not compile `playground` (lib) due to 1 previous error
```

### Rust Version

```Shell
rustc 1.75.0
```


### Anything else?

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b28797586ea1abf6da072badb426608c)

<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"veera-sivarajan"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -1247,10 +1247,23 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
             (&ty::Adt(def1, sub1), &ty::Adt(def2, sub2)) => {
                 let did1 = def1.did();
                 let did2 = def2.did();
-                let sub_no_defaults_1 =
-                    self.tcx.generics_of(did1).own_args_no_defaults(self.tcx, sub1);
-                let sub_no_defaults_2 =
-                    self.tcx.generics_of(did2).own_args_no_defaults(self.tcx, sub2);
+
+                let generics1 = self.tcx.generics_of(did1);
+                let generics2 = self.tcx.generics_of(did2);
+
+                let non_default_after_default = generics1
+                    .check_concrete_type_after_default(self.tcx, sub1)
+                    || generics2.check_concrete_type_after_default(self.tcx, sub2);
+                let sub_no_defaults_1 = if non_default_after_default {
+                    generics1.own_args(sub1)
+                } else {
+                    generics1.own_args_no_defaults(self.tcx, sub1)
+                };
+                let sub_no_defaults_2 = if non_default_after_default {
+                    generics2.own_args(sub2)
+                } else {
+                    generics2.own_args_no_defaults(self.tcx, sub2)
+                };
                 let mut values = (DiagStyledString::new(), DiagStyledString::new());
                 let path1 = self.tcx.def_path_str(did1);
                 let path2 = self.tcx.def_path_str(did2);

--- a/compiler/rustc_middle/src/ty/generics.rs
+++ b/compiler/rustc_middle/src/ty/generics.rs
@@ -360,6 +360,33 @@ impl<'tcx> Generics {
         let own = &args[self.parent_count..][..self.params.len()];
         if self.has_self && self.parent.is_none() { &own[1..] } else { own }
     }
+
+    /// Returns true if a concrete type is specified after a default type.
+    /// For example, consider `struct T<W = usize, X = Vec<W>>(W, X)`
+    /// `T<usize, String>` will return true
+    /// `T<usize>` will return false
+    pub fn check_concrete_type_after_default(
+        &'tcx self,
+        tcx: TyCtxt<'tcx>,
+        args: &'tcx [ty::GenericArg<'tcx>],
+    ) -> bool {
+        let mut default_param_seen = false;
+        for param in self.params.iter() {
+            if param
+                .default_value(tcx)
+                .is_some_and(|default| default.instantiate(tcx, args) == args[param.index as usize])
+            {
+                default_param_seen = true;
+            } else if default_param_seen
+                && param.default_value(tcx).is_some_and(|default| {
+                    default.instantiate(tcx, args) != args[param.index as usize]
+                })
+            {
+                return true;
+            }
+        }
+        false
+    }
 }
 
 /// Bounds on generics.

--- a/compiler/rustc_middle/src/ty/generics.rs
+++ b/compiler/rustc_middle/src/ty/generics.rs
@@ -372,17 +372,14 @@ impl<'tcx> Generics {
     ) -> bool {
         let mut default_param_seen = false;
         for param in self.params.iter() {
-            if param
-                .default_value(tcx)
-                .is_some_and(|default| default.instantiate(tcx, args) == args[param.index as usize])
+            if let Some(inst) =
+                param.default_value(tcx).map(|default| default.instantiate(tcx, args))
             {
-                default_param_seen = true;
-            } else if default_param_seen
-                && param.default_value(tcx).is_some_and(|default| {
-                    default.instantiate(tcx, args) != args[param.index as usize]
-                })
-            {
-                return true;
+                if inst == args[param.index as usize] {
+                    default_param_seen = true;
+                } else if default_param_seen {
+                    return true;
+                }
             }
         }
         false

--- a/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.rs
+++ b/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.rs
@@ -5,7 +5,7 @@ fn main() {
     let c: What<usize, String> = What(1, String::from("meow"));
     b = c; //~ ERROR mismatched types
 
-    let mut e: What<usize> = What(5, vec![1, 2, 3]);
-    let f: What<usize, Vec<String>> = What(1, vec![String::from("meow")]);
-    e = f; //~ ERROR mismatched types
+    let mut f: What<usize, Vec<String>> = What(1, vec![String::from("meow")]);
+    let e: What<usize> = What(5, vec![1, 2, 3]);
+    f = e; //~ ERROR mismatched types
 }

--- a/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.rs
+++ b/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.rs
@@ -1,0 +1,11 @@
+struct What<W = usize, X = Vec<W>>(W, X);
+
+fn main() {
+    let mut b: What<usize> = What(5, vec![1, 2, 3]);
+    let c: What<usize, String> = What(1, String::from("meow"));
+    b = c; //~ ERROR mismatched types
+
+    let mut e: What<usize> = What(5, vec![1, 2, 3]);
+    let f: What<usize, Vec<String>> = What(1, vec![String::from("meow")]);
+    e = f; //~ ERROR mismatched types
+}

--- a/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.stderr
+++ b/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.stderr
@@ -13,14 +13,14 @@ LL |     b = c;
 error[E0308]: mismatched types
   --> $DIR/clarify-error-for-generics-with-default-issue-120785.rs:10:9
    |
-LL |     let mut e: What<usize> = What(5, vec![1, 2, 3]);
-   |                ----------- expected due to this type
-LL |     let f: What<usize, Vec<String>> = What(1, vec![String::from("meow")]);
-LL |     e = f;
-   |         ^ expected `What`, found `What<usize, Vec<String>>`
+LL |     let mut f: What<usize, Vec<String>> = What(1, vec![String::from("meow")]);
+   |                ------------------------ expected due to this type
+LL |     let e: What<usize> = What(5, vec![1, 2, 3]);
+LL |     f = e;
+   |         ^ expected `What<usize, Vec<String>>`, found `What`
    |
-   = note: expected struct `What<_, Vec<usize>>`
-              found struct `What<_, Vec<String>>`
+   = note: expected struct `What<_, Vec<String>>`
+              found struct `What<_, Vec<usize>>`
 
 error: aborting due to 2 previous errors
 

--- a/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.stderr
+++ b/tests/ui/type/clarify-error-for-generics-with-default-issue-120785.stderr
@@ -1,0 +1,27 @@
+error[E0308]: mismatched types
+  --> $DIR/clarify-error-for-generics-with-default-issue-120785.rs:6:9
+   |
+LL |     let mut b: What<usize> = What(5, vec![1, 2, 3]);
+   |                ----------- expected due to this type
+LL |     let c: What<usize, String> = What(1, String::from("meow"));
+LL |     b = c;
+   |         ^ expected `What`, found `What<usize, String>`
+   |
+   = note: expected struct `What<_, Vec<usize>>`
+              found struct `What<_, String>`
+
+error[E0308]: mismatched types
+  --> $DIR/clarify-error-for-generics-with-default-issue-120785.rs:10:9
+   |
+LL |     let mut e: What<usize> = What(5, vec![1, 2, 3]);
+   |                ----------- expected due to this type
+LL |     let f: What<usize, Vec<String>> = What(1, vec![String::from("meow")]);
+LL |     e = f;
+   |         ^ expected `What`, found `What<usize, Vec<String>>`
+   |
+   = note: expected struct `What<_, Vec<usize>>`
+              found struct `What<_, Vec<String>>`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0308`.

